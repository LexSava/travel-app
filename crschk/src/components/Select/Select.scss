@import '../../variables.scss';

.language-select {
  border-radius: 18px;
  height: 36px;
  padding: 0 15px;
  width: 150px;
  border: none;
  margin-right: 10px;
  background-color: $color-focus;
  color: white;
  font-size: 18px;

  &:hover {
    filter: brightness(1.2);
    cursor: pointer;
  }

  @media screen and (max-width: 1400px) {
    margin-right: 0;
  }
}