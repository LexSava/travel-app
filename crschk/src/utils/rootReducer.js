import actions from "./actions";

const getData = (key, initialValue) => {
  const data = localStorage.getItem(key);

  if (!data) {
    localStorage.setItem(key, initialValue);
  }

  let result = data;
  try {
    result = JSON.parse(data);
  } catch {
    result = data;
  }

  return result ? result : initialValue;
};

export const initialState = {
  locale: getData("lang", "ru_RU"),
  dict: {},
  user: getData("user", null),
  searchValue: "",
  authForm: {
    isFormOpen: false,
    isSignup: true,
  },
};

export default function rootReducer(state = initialState, action) {
  switch (action.type) {
    case actions.SET_LOCALE: {
      localStorage.setItem("lang", action.payload);
      return {
        ...state,
        locale: action.payload,
      };
    }
    case actions.ADD_LOCALE: {
      return {
        ...state,
        dict: action.payload,
      };
    }
    case actions.SET_USER: {
      localStorage.setItem("user", JSON.stringify(action.user));
      return {
        ...state,
        user: getData("user", null),
      };
    }
    case actions.REMOVE_USER: {
      localStorage.setItem("user", JSON.stringify(null));
      return {
        ...state,
        user: getData("user", null),
      };
    }
    case actions.SEARCH: {
      return {
        ...state,
        searchValue: action.payload,
      };
    }
    case actions.SET_AUTHFORM: {
      return {
        ...state,
        authForm: {
          ...state.authForm,
          ...action.payload,
        },
      };
    }
    default:
      return state;
  }
}
