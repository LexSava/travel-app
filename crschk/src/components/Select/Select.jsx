import React, { useEffect, useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import actions from '../../utils/actions';
import { getLocales } from '../../utils/api';

import './Select.scss';

export default function Select() {
  const [locales, setLocales] = useState(null);
  const { locale } = useSelector(state => state);

  const dispatch = useDispatch();

  useEffect(() => {
    getLocales()
      .then(setLocales)
  }, [])

  const setLanguage = (e) => {
    dispatch({type: actions.SET_LOCALE, payload: e.target.value})
  }

  return (
    <select className="language-select" value={locale} onChange={setLanguage}>
      { locales &&
        locales.map((lang, i) => (
          <option key={i} value={lang.lang}>{lang.name}</option>
        ))
      }
    </select>
  )
}