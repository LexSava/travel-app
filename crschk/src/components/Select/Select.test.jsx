import React from "react";
import { Provider } from "react-redux";
import renderer, { act } from "react-test-renderer";
global.fetch = require("node-fetch");
import rootReducer from "../../utils/rootReducer";
import "babel-polyfill";

import Select from "./Select";
import { createStore } from "redux";

describe("Search", () => {
  const renderComp = () =>
    renderer.create(
      <Provider store={store}>
        <Select />
      </Provider>
    );
  const setState = jest.fn();
  const useStateMock = (initState) => [initState, setState];

  afterEach(() => {
    jest.clearAllMocks();
  });
  const store = createStore(rootReducer);

  it("Should render search with img", () => {
    const tree = renderComp().toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("Should set state on select change", async () => {
    jest.spyOn(React, "useState").mockImplementation(useStateMock);
    const wrapper = renderComp();
    const select = wrapper.root.findByType("select");
    act(() => select.props.onChange({ target: { value: "test" } }));
    expect(store.getState().locale).toBe("test");
  });
});
