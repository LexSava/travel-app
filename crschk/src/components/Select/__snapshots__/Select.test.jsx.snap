// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search Should render search with img 1`] = `
<select
  className="language-select"
  onChange={[Function]}
  value="ru_RU"
/>
`;
